libsolutil/Algorithms.h

```diff
@@ -104,13 +103,14 @@ struct BreadthFirstSearch
 	{
 		while (!verticesToTraverse.empty())
 		{
-			V v = *verticesToTraverse.begin();
-			verticesToTraverse.erase(verticesToTraverse.begin());
+			V v = std::move(verticesToTraverse.front());
+			verticesToTraverse.pop_front();
+
 			visited.insert(v);
```

Comment:
We only mark a node as visited when we visit it. If multiple edges point to the same node, then this node can be inserted into the queue multiple times and thus will be visited multiple times. We should at least change this line to
```
if (!visited.insert(v).second)
  continue;
```
here.

Reply:
Another solution would be to change the `visited` set into a `queued` set that is set to true once the node has been seen in line 113

Reply:
It cannot be inserted into the queue multiple times because exactly because we check before adding to the queue

Reply:
The only case where that's possible is if you initialize `verticesToVisit` with multiple copies of the same value